com.unity.netcode.gameobjects/Prototyping/Interpolator/BufferedLinearInterpolator.cs

```diff
@@ -0,0 +1,232 @@
+            {
+                double range = FixedOrTime(m_EndTimeConsumed) - FixedOrTime(m_StartTimeConsumed);
+                float t;
+                if (range == 0)
+                {
+                    t = 1;
+                }
+                else
+                {
+                    t = (float) ((RenderTime - FixedOrTime(m_StartTimeConsumed)) / range);
+                }
+
+                if (t > 3) // max extrapolation
+                {
+                    t = 1;
+                }
+
+                if (Debug.isDebugBuild)
+                {
+                    Debug.Assert(t >= 0, $"t must be bigger or equal than 0. range {range}, RenderTime {RenderTime}, Start time {FixedOrTime(m_StartTimeConsumed)}, end time {FixedOrTime(m_EndTimeConsumed)}");
+                }
+
+                SimpleInterpolator.AddMeasurement(Interpolate(m_InterpStartValue, m_InterpEndValue, t), default); // using simple interpolation so there's no jump
+                m_CurrentInterpValue = SimpleInterpolator.Update(deltaTime);
+            }
+
+            return m_CurrentInterpValue;
+        }
+
+        public void FixedUpdate(float fixedDeltaTime)
+        {
+        }
+
+        public void AddMeasurement(T newMeasurement, NetworkTime sentTime)
+        {
+            if (m_Buffer.Count >= k_BufferSizeLimit)
+            {
+                Debug.LogWarning("Going over buffer size limit while adding new interpolation values, interpolation buffering isn't consuming fast enough, removing oldest value now.");
+                m_Buffer.RemoveAt(m_Buffer.Count - 1);
+            }
+
+            if (sentTime.Time > m_EndTimeConsumed.Time || m_LifetimeConsumedCount == 0) // treat only if value is newer than the one being interpolated to right now
+            {
+                m_Buffer.Add(new BufferedItem {item = newMeasurement, timeSent = sentTime});
+                m_Buffer.Sort((item1, item2) => item2.timeSent.Time.CompareTo(item1.timeSent.Time));
+            }
+        }
+
+        public T GetInterpolatedValue()
+        {
+            return m_CurrentInterpValue;
+        }
+
+        public void OnDestroy()
+        {
+        }
+
+        protected abstract T Interpolate(T start, T end, float time);
+
+        protected abstract SimpleInterpolator<T> SimpleInterpolator { get; }
+    }
+
+    public class BufferedLinearInterpolatorFloat : BufferedLinearInterpolator<float>
+    {
+        protected override float Interpolate(float start, float end, float time)
+        {
+            return Mathf.LerpUnclamped(start, end, time);
+        }
+
+        protected override SimpleInterpolator<float> SimpleInterpolator { get; } = new SimpleInterpolatorFloat();
+    }
+
+    public class BufferedLinearInterpolatorFloatForScale : BufferedLinearInterpolatorFloat
+    {
+        protected override float Interpolate(float start, float end, float time)
+        {
+            // scale can't be negative, stopping negative extrapolation
```

Comment:
I can make scale negative in the Unity inspector? It is not recommended for static batching and stuff but technically you can do it.

Reply:
negative scaling makes the model mirrored. -1 on scale z is a technique to mirror models such as first-person hands (left to right and vice-versa, for instance: Counter-Strike weapon viewmodels)

Reply:
mmm that's true... I'll investigate why the editor was complaining about negative scale when I tested this...

Reply:
haaa, I read too quickly. It was a "BoxColliders does not support negative scale or size." warning
Thanks for catching this, will remove this check